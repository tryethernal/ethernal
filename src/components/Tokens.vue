<template>
    <v-container fluid>
        <v-data-table
            :loading="loading"
            :items="tokens"
            :headers="headers"
            :sort-by="currentOptions.sortBy[0]"
            :must-sort="true"
            :sort-desc="true"
            :server-items-length="tokenCount"
            :footer-props="{
                itemsPerPageOptions: [10, 25, 100]
            }"
            item-key="address"
            @update:options="getTokens">
            <template v-slot:item.address="{ item }">
                <Hash-Link :type="'address'" :hash="item.address" />
            </template>
            <template v-slot:item.contractName="{ item }">
                {{ item.name }}
            </template>
            <template v-slot:item.tokenName="{ item }">
                {{ item.tokenName }}
            </template>
            <template v-slot:item.tokenSymbol="{ item }">
                {{ item.tokenSymbol }}
            </template>
            <template v-slot:item.timestamp="{ item }">
                <span v-if="item.timestamp">{{ parseInt(item.timestamp) | moment('YYYY-MM-DD h:mm:ss A') }}</span>
            </template>
            <template v-slot:item.tags="{ item }">
                <v-chip v-for="(pattern, idx) in item.patterns" :key="idx" x-small class="success mr-2">
                    {{ formatContractPattern(pattern) }}
                </v-chip>
            </template>
        </v-data-table>
    </v-container>
</template>
<script>
import HashLink from '@/components/HashLink';
import { formatContractPattern } from '@/lib/utils';

export default {
    name: 'Contracts',
    components: {
        HashLink
    },
    data: () => ({
        loading: true,
        tokens: [],
        tokenCount: 0,
        headers: [
            {
                text: 'Address',
                value: 'address'
            },
            {
                text: 'Contract Name',
                value: 'contractName'
            },
            {
                text: 'Token Name',
                value: 'tokenName'
            },
            {
                text: 'Token Symbol',
                value: 'tokenSymbol'
            },
            {
                text: 'Deployed On',
                value: 'timestamp'
            },
            {
                text: '',
                value: 'tags'
            }
        ],
        currentOptions: { page: 1, itemsPerPage: 10, sortBy: ['timestamp'], sortDesc: [true] }
    }),
    mounted: function() {
        this.pusher.onNewToken(() => this.getTokens(this.currentOptions), this);
    },
    methods: {
        getTokens: function(newOptions) {
            this.loading = true;

            if (newOptions)
                this.currentOptions = newOptions;

            const options = {
                page: this.currentOptions.page,
                itemsPerPage: this.currentOptions.itemsPerPage,
                order: this.currentOptions.sortDesc[0] === false ? 'asc' : 'desc',
                onlyTokens: true
            };

            this.server.getContracts(options)
                .then(({ data }) => {
                    this.tokens = data.items;
                    this.tokenCount = data.total;
                })
                .catch(console.log)
                .finally(() => this.loading = false);
        },
        formatContractPattern
    }
}
</script>
