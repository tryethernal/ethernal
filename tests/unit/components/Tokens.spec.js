import flushPromises from 'flush-promises';
import MockHelper from '../MockHelper';

import Tokens from '@/components/Tokens.vue';

describe('Tokens.vue', () => {
    let helper;

    beforeEach(() => {
        helper = new MockHelper();
    });

    it('Should display token contracts', async () => {
        jest.spyOn(helper.mocks.server, 'getContracts')
            .mockResolvedValue({
                data: {
                    items: [
                        { timestamp: '1636557049', address: '0x123', contractName: 'Ethernal Token', tokenName: 'Ethernal', tokenSymbol: 'ETL', tokenDecimals: 18, patterns: ['erc20'] },
                        { timestamp: '1636557049', address: '0x124', contractName: 'USD Coin', tokenName: 'USDC', tokenSymbol: 'USDC', tokenDecimals: 6, patterns: ['erc20', 'proxy'] }
                    ]
                }
            });

        const wrapper = helper.mountFn(Tokens, {
            stubs: ['Hash-Link']
        });
        await new Promise(process.nextTick);

        expect(wrapper.html()).toMatchSnapshot();
    });
});
