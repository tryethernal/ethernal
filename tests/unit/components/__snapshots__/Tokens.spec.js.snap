// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokens.vue Should display token contracts 1`] = `
<div class="container container--fluid">
  <div class="v-data-table v-data-table--has-bottom theme--light">
    <div class="v-data-table__wrapper">
      <table>
        <colgroup>
          <col class="">
          <col class="">
          <col class="">
          <col class="">
        </colgroup>
        <thead class="v-data-table-header">
          <tr>
            <th role="columnheader" scope="col" aria-label="Address: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Address</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Name: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Name</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Symbol: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Symbol</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span></span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          </tr>
        </thead>
        <tbody>
          <tr class="">
            <td class="text-start">
              <hash-link-stub hash="0x123" type="token"></hash-link-stub>
            </td>
            <td class="text-start">
              Ethernal
            </td>
            <td class="text-start">
              ETL
            </td>
            <td class="text-start"><span class="success mr-2 v-chip v-chip--no-color theme--light v-size--x-small"><span class="v-chip__content">
                ERC 20
            </span></span></td>
          </tr>
          <tr class="">
            <td class="text-start">
              <hash-link-stub hash="0x124" type="token"></hash-link-stub>
            </td>
            <td class="text-start">
              USDC
            </td>
            <td class="text-start">
              USDC
            </td>
            <td class="text-start"><span class="success mr-2 v-chip v-chip--no-color theme--light v-size--x-small"><span class="v-chip__content">
                ERC 20
            </span></span><span class="success mr-2 v-chip v-chip--no-color theme--light v-size--x-small"><span class="v-chip__content">
                Proxy
            </span></span></td>
          </tr>
        </tbody>
      </table>
    </div>
    <div class="v-data-footer">
      <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-15" class="v-input__slot">
              <div class="v-select__slot">
                <div class="v-select__selections">
                  <div class="v-select__selection v-select__selection--comma">10</div><input aria-label="Rows per page:" id="input-15" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                </div><input type="hidden" value="10">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-data-footer__pagination">1-2 of 2</div>
      <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div>
      <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button></div>
    </div>
  </div>
</div>
`;
